magnum capi changes (#557)
stackhpc/stackhpc-kayobe-config: doc/source/configuration/magnum-capi.rst

=========================
Magnum Cluster API Driver
=========================

> 

Reviewer: Intro:
what is the new driver?
how does it relate to the old driver?
can I use both together?

Author: are changes sufficient?

Reviewer: LGTM, thanks

> Prequisites for deploying the CAPI driver in magnum:

Reviewer: ```suggestion
Prerequisites for deploying the CAPI driver in magnum:
```

Management Cluster
==================

> The CAPI driver relies on a management cluster to manage tenant kubernetes clusters. 

Reviewer 2: Can you please note that this is its self a Kubernetes cluster installed using cluster API, running inside the OpenStack cloud.

> The CAPI driver relies on a management cluster to manage tenant kubernetes clusters.

Reviewer: ```suggestion
The CAPI driver relies on a management Kubernetes cluster to manage tenant Kubernetes clusters.
```

> The easiest way to get one is by deploying [this](https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config) branch of azimuth-config, and look at the `capi-mgmt-example` environment.

Reviewer: Will this always be on a feature branch?

Author: I don't think so but it is at the moment 😁

> The easiest way to get one is by deploying [this](https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config) branch of azimuth-config, and look at the `capi-mgmt-example` environment.

Reviewer: This could use some more pointers for those unfamiliar with Azimuth deployment.

Author: done?

> Ensure that you have set `capi_cluster_apiserver_floating_ip: true`, as the management cluster will need an externally accessible IP.

Reviewer 2: Can we define external here please? Who need to talk to the end points we have here? Maybe reference the main azimuth docs here?

> 

Reviewer 2: Can we note the example .sh file that will get setup on the Kayobe seed node so you can quickly access the bin/seed-ssh script and the monitoring port forward scripts that we have explain here:
https://stackhpc.github.io/azimuth-config/deployment/
and here:
https://stackhpc.github.io/azimuth-config/debugging/access-ha/

Maybe note that generally we expect most day 2 ops to be done via CD pipelines?

Kayobe Config
==============
Ensure that your kayobe-config branch is up to date on stackhpc/yoga.

> Copy the kubeconfig found at `kubeconfig-capi-mgmt-<your-environment>.yaml` to your kayobe environment (e.g. `<your-environment>/kolla/config/magnum/kubeconfig`.

Reviewer: Are these two `your-environments` referring to the same trhing?

> Ensure that your magnum.conf has the following set: 

Reviewer 2: Can we please note why, i.e. this is used to generate the application credentials config injected into the Kubernetes clusters, such that it is usable from within an OpenStack project, so you can't use the "internal API" end point here.

```
[nova_client]
endpoint_type = publicURL
```

Control Plane
==============

> Ensure that the nodes (either controllers or dedicated network hosts) that you are running the magnum containers on have internet connectivity (so that the magnum containers can reach the IP listed in the kubeconfig).

Reviewer: Do you mean the actual Internet or just the network on which the management cluster API is available on?

> Ensure that the nodes (either control plane nodes or dedicated network nodes) that you are running the magnum containers on have internet connectivity (so that the magnum containers can reach the IP listed in the kubeconfig). 

Reviewer 3: ```suggestion
Ensure that the hosts (either controllers or dedicated network hosts) that you are running the magnum containers on have internet connectivity (so that the magnum containers can reach the IP listed in the kubeconfig). 
```

Magnum Templates
===============

> (openstack-config reference templates to be added shortly)

Reviewer 2: Can we link to the CI we do on the helm charts, and where the latest images and tags are defined in the azimuth-config repo? Maybe that is overkill I guess?

=========================
Magnum Cluster API Driver
=========================
A new driver for magnum has been written. It is an alternative to heat (as heat gets phased out due to maintenance burden) that allows the definition of clusters as Kubernetes CRDs as opposed to heat templates. The two are compatible and can both be active on the same deployment, and the decision of which driver is used for a given template depends on certain parameters inferred from the template. For the new driver, these are `{'server_type' : 'vm', 'os' : 'ubuntu', 'coe': kubernetes'}`.
Drivers can be enabled and disabled via the `disabled_drivers` parameter of `[drivers]` under `magnum.conf`.

Prerequisites for deploying the CAPI driver in magnum:

Management Cluster
===================
The CAPI driver relies on a management Kubernetes cluster, installed inside the cloud, to manage tenant Kubernetes clusters.

> The easiest way to get one is by deploying [this](https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config) branch of azimuth-config, and look at the `capi-mgmt-example` environment. Refer to the [azimuth-config wiki](https://stackhpc.github.io/azimuth-config/) for detailed steps on how to deploy.

Reviewer 4: ```suggestion
The easiest way to get one is by deploying `this <https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config>`__ branch of azimuth-config, and look at the `capi-mgmt-example` environment. Refer to the `azimuth-config wiki <https://stackhpc.github.io/azimuth-config/>`__ for detailed steps on how to deploy.
```

> There is a helper script for setting up your environment to run manual commands at `tools/azimuth-config-dev.sh`. The usage is : `./tools/azimuth-config-dev.sh <environment`. It assumes you follow the naming convention of `<environment>-azimuth-config` for your fork, as well as `environments/<environment>` for your mix-in environment.

Reviewer 4: ```suggestion
There is a helper script for setting up your environment to run manual commands at `tools/azimuth-config-dev.sh`. The usage is : `./tools/azimuth-config-dev.sh <environment>`. It assumes you follow the naming convention of `<environment>-azimuth-config` for your fork, as well as `environments/<environment>` for your mix-in environment.
```

Ensure that you have set `capi_cluster_apiserver_floating_ip: true`, as the management cluster will need an externally accessible IP. The external network this corresponds to is whatever you have set `azimuth_capi_operator_external_network_id` to. This network needs to be reachable from wherever the magnum container is running.

> It's preferrable that most Day 2 ops be done via a [CD Pipeline](https://stackhpc.github.io/azimuth-config/deployment/automation/).

Reviewer 4: ```suggestion
It's preferable that most Day 2 ops be done via a `CD Pipeline <https://stackhpc.github.io/azimuth-config/deployment/automation/>`__.
```

Kayobe Config
==============
Ensure that your kayobe-config branch is up to date on stackhpc/yoga.

There is a helper script for setting up your environment, located at `tools/kayobe-dev.sh`. Again, the usage is `./tools/kayobe-dev.sh <environment>` where this is your kayobe environment located at `etc/kayobe/environments/<environment>`.

Copy the kubeconfig found at `kubeconfig-capi-mgmt-<your-az-environment>.yaml` to your kayobe environment (e.g. `<your-skc-environment>/kolla/config/magnum/kubeconfig`.

Ensure that your magnum.conf has the following set:

> ```
> [nova_client]
> endpoint_type = publicURL
> ```

Reviewer 4: Triple backticks don't correctly show a code block in .rst files
```suggestion
.. code-block:: yaml

    [nova_client]
    endpoint_type = publicURL

```

This is used to generate the application credential config injected into the tenant Kubernetes clusters, such that it is usable from within an OpenStack project, so you can't use the "internal API" end point here.

Control Plane
==============
Ensure that the nodes (either controllers or dedicated network hosts) that you are running the magnum containers on have connectivity to the network on which your management cluster has a floating IP (so that the magnum containers can reach the IP listed in the kubeconfig).

Magnum Templates
================

> [azimuth-images](https://github.com/stackhpc/azimuth-images) builds the required Ubuntu Kubernetes images, and [capi-helm-charts](https://github.com/stackhpc/capi-helm-charts/blob/main/.github/workflows/test.yaml) CI runs conformance tests on each image built.

Reviewer 4: ```suggestion
`azimuth-images <https://github.com/stackhpc/azimuth-images>`__ builds the required Ubuntu Kubernetes images, and `capi-helm-charts <https://github.com/stackhpc/capi-helm-charts/blob/main/.github/workflows/test.yaml>`__ CI runs conformance tests on each image built.
```

> Magnum templates can be deployed using [openstack-config](https://github.com/stackhpc/openstack-config). Typically, you would create a fork `<environment>-config` of this repository, move the resources defined in `examples/capi-templates-images.yml` into `etc/openstack-config/openstack-config.yml`, and then follow the instructions in the readme to deploy these.

Reviewer 4: ```suggestion
Magnum templates can be deployed using `openstack-config <https://github.com/stackhpc/openstack-config>`__. Typically, you would create a fork `<environment>-config` of this repository, move the resources defined in `examples/capi-templates-images.yml` into `etc/openstack-config/openstack-config.yml`, and then follow the instructions in the readme to deploy these.
```

Magnum Cluster API Driver
=========================
A new driver for magnum has been written. It is an alternative to heat (as heat gets phased out due to maintenance burden) that allows the definition of clusters as Kubernetes CRDs as opposed to heat templates. The two are compatible and can both be active on the same deployment, and the decision of which driver is used for a given template depends on certain parameters inferred from the template. For the new driver, these are `{'server_type' : 'vm', 'os' : 'ubuntu', 'coe': kubernetes'}`.
Drivers can be enabled and disabled via the `disabled_drivers` parameter of `[drivers]` under `magnum.conf`.

Prerequisites for deploying the CAPI driver in magnum:

Management Cluster
===================
The CAPI driver relies on a management Kubernetes cluster, installed inside the cloud, to manage tenant Kubernetes clusters.
The easiest way to get one is by deploying `this <https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config>`__ branch of azimuth-config, and look at the `capi-mgmt-example` environment. Refer to the `azimuth-config wiki <https://stackhpc.github.io/azimuth-config/>`__ for detailed steps on how to deploy.

> There is a helper script for setting up your environment to run manual commands at `tools/azimuth-config-dev.sh`. The usage is : `./tools/azimuth-config-dev.sh <environment>`. It assumes you follow the naming convention of `<environment>-azimuth-config` for your fork, as well as `environments/<environment>` for your mix-in environment.

Reviewer 4: ```suggestion
There is a helper script for setting up your environment to run manual commands at `tools/azimuth-config-dev.sh`. The usage is: `./tools/azimuth-config-dev.sh <environment>`. It assumes you follow the naming convention of `<environment>-azimuth-config` for your fork, as well as `environments/<environment>` for your mix-in environment.
```

Ensure that you have set `capi_cluster_apiserver_floating_ip: true`, as the management cluster will need an externally accessible IP. The external network this corresponds to is whatever you have set `azimuth_capi_operator_external_network_id` to. This network needs to be reachable from wherever the magnum container is running.

It's preferable that most Day 2 ops be done via a `CD Pipeline <https://stackhpc.github.io/azimuth-config/deployment/automation/>`__.

Kayobe Config
==============
Ensure that your kayobe-config branch is up to date on stackhpc/yoga.

There is a helper script for setting up your environment, located at `tools/kayobe-dev.sh`. Again, the usage is `./tools/kayobe-dev.sh <environment>` where this is your kayobe environment located at `etc/kayobe/environments/<environment>`.

Copy the kubeconfig found at `kubeconfig-capi-mgmt-<your-az-environment>.yaml` to your kayobe environment (e.g. `<your-skc-environment>/kolla/config/magnum/kubeconfig`.

> Ensure that your magnum.conf has the following set:
> .. code-block:: yaml

Reviewer 4: ```suggestion
Ensure that your magnum.conf has the following set:

.. code-block:: yaml
```

Magnum Cluster API Driver
=========================
A new driver for magnum has been written. It is an alternative to heat (as heat gets phased out due to maintenance burden) that allows the definition of clusters as Kubernetes CRDs as opposed to heat templates. The two are compatible and can both be active on the same deployment, and the decision of which driver is used for a given template depends on certain parameters inferred from the template. For the new driver, these are `{'server_type' : 'vm', 'os' : 'ubuntu', 'coe': kubernetes'}`.
Drivers can be enabled and disabled via the `disabled_drivers` parameter of `[drivers]` under `magnum.conf`.

Prerequisites for deploying the CAPI driver in magnum:

Management Cluster
===================
The CAPI driver relies on a management Kubernetes cluster, installed inside the cloud, to manage tenant Kubernetes clusters.
The easiest way to get one is by deploying `this <https://github.com/stackhpc/azimuth-config/tree/feature/capi-mgmt-config>`__ branch of azimuth-config, and look at the `capi-mgmt-example` environment. Refer to the `azimuth-config wiki <https://stackhpc.github.io/azimuth-config/>`__ for detailed steps on how to deploy.

> There is a helper script for setting up your environment to run manual commands at `tools/azimuth-config-dev.sh`. The usage is: `./tools/azimuth-config-dev.sh <environment>`. It assumes you follow the naming convention of `<environment>-azimuth-config` for your fork, as well as `environments/<environment>` for your mix-in environment.

Reviewer: Still referencing the helper scripts

Ensure that you have set `capi_cluster_apiserver_floating_ip: true`, as the management cluster will need an externally accessible IP. The external network this corresponds to is whatever you have set `azimuth_capi_operator_external_network_id` to. This network needs to be reachable from wherever the magnum container is running.

It's preferable that most Day 2 ops be done via a `CD Pipeline <https://stackhpc.github.io/azimuth-config/deployment/automation/>`__.

Kayobe Config
==============

> Ensure that your kayobe-config branch is up to date on stackhpc/yoga.

Reviewer: ```suggestion
Ensure that your kayobe-config branch is up to date on |current_release_git_branch_name|.
```

Author: it's only on yoga at the moment

Reviewer: Yes, but it will be on zed when we create that soon. This variable avoids needing to change the branch each time we move to a new release.
